modules/transactionpool/update.go

```diff
@@ -8,6 +8,124 @@ import (
+		for exists {
+			ret = next
+			forwards[prev] = next // collapse the forwards function to prevent quadratic runtime of findSets.
+			next, exists = forwards[next]
+		}
+		return ret
+	}
+
+	// Add the transactions to the setup one-by-one, merging them as they belong
+	// to a set.
+	for i, t := range ts {
+		// Check if the inputs depend on any previous transaction outputs.
+		tid := t.ID()
+		parentSets := make(map[int]struct{})
+		for _, obj := range t.SiacoinInputs {
+			txid, exists := objMap[ObjectID(obj.ParentID)]
+			if exists {
+				parentSet := forward(txMap[txid])
+				parentSets[parentSet] = struct{}{}
+			}
+		}
+		for _, obj := range t.FileContractRevisions {
+			txid, exists := objMap[ObjectID(obj.ParentID)]
+			if exists {
+				parentSet := forward(txMap[txid])
+				parentSets[parentSet] = struct{}{}
+			}
+		}
+		for _, obj := range t.StorageProofs {
+			txid, exists := objMap[ObjectID(obj.ParentID)]
+			if exists {
+				parentSet := forward(txMap[txid])
+				parentSets[parentSet] = struct{}{}
+			}
+		}
+		for _, obj := range t.SiafundInputs {
+			txid, exists := objMap[ObjectID(obj.ParentID)]
+			if exists {
+				parentSet := forward(txMap[txid])
+				parentSets[parentSet] = struct{}{}
+			}
+		}
+
+		// Determine the new counter for this transaction.
+		if len(parentSets) == 0 {
+			// No parent sets. Make a new set for this transaction.
+			txMap[tid] = i
+			setMap[i] = []types.Transaction{t}
+			// Don't need to add anything for the file contract outputs, storage
+			// proof outputs, siafund claim outputs; these outputs are not
+			// allowed to be spent until 50 confirmations.
+		} else {
+			// There are parent sets, pick one as the base and then merge the
+			// rest into it.
+			parentsSlice := make([]int, 0, len(parentSets))
+			for j := range parentSets {
+				parentsSlice = append(parentsSlice, j)
+			}
+			base := parentsSlice[0]
+			txMap[tid] = base
+			for _, j := range parentsSlice[1:] {
+				// Forward any future transactions pointing at this set to the
+				// base set.
+				forwards[j] = base
+				// Combine the transactions in this set with the transactions in
+				// the base set.
+				setMap[base] = append(setMap[base], setMap[j]...)
+				// Delete this set map, it has been merged with the base set.
+				delete(setMap, j)
+			}
+			// Add this transaction to the base set.
+			setMap[base] = append(setMap[base], t)
+		}
+
+		// Mark this transaction's outputs as potential inputs to future
+		// transactions.
+		for j := range t.SiacoinOutputs {
+			scoid := t.SiacoinOutputID(uint64(j))
+			objMap[ObjectID(scoid)] = tid
+		}
+		for j := range t.FileContracts {
+			fcid := t.FileContractID(uint64(j))
+			objMap[ObjectID(fcid)] = tid
+		}
+		for j := range t.FileContractRevisions {
+			fcid := t.FileContractRevisions[j].ParentID
+			objMap[ObjectID(fcid)] = tid
+		}
+		for j := range t.SiafundOutputs {
+			sfoid := t.SiafundOutputID(uint64(j))
+			objMap[ObjectID(sfoid)] = tid
+		}
```

Comment:
ok, wasn't sure because of this comment:
>```
>// Don't need to add anything for the file contract outputs, storage
>// proof outputs, siafund claim outputs; these outputs are not
>// allowed to be spent until 50 confirmations.
>```